src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -755,6 +794,33 @@ public List<ActivationAddress> GetLocalCacheData(GrainId grain)
                 null;
         }
 
+        public Task<AddressesAndTag> LookupInCluster(GrainId grainId, string clusterId)
+        {
+            if (clusterId == null)
+                throw new ArgumentNullException("clusterId");
+
+            if (clusterId == ClusterId)
+            {
+                return LookupAsync(grainId);
+            }
+            else
+            {
+                // find gateway
+                var gossipOracle = Silo.CurrentSilo.LocalMultiClusterOracle;
+                var clusterGatewayAddress = gossipOracle.GetRandomClusterGateway(clusterId);
```

Comment:
`GetRandomClusterGateway` executes an expensive LINQ expression upon every call. I think we need either to cache (and expire) values here or optimize implementation of `GetRandomClusterGateway` to cache the list of gateways and rebuild it only when the set of gateways changes or both.